### PR TITLE
Support !JitDoOldStructRetyping on  other platforms.

### PR DESCRIPTION
**Pull request by sandreenko**
Add `JitDoOldStructRetyping=0` support for x86 (linux and windows) and x64 Unix (windows was already supported).
For x86 the only change was to move the code out of `!FEATURE_MULTIREG` and keep other platforms working. The only multireg there are long, not structs.
For x64 Unix it was more interesting: it supports passing 3, 5, 6, 7 bytes structs in registers and it supports passing fields in float registers.

Based on logic for 3, 5, 6, 7 I noticed that we do widening for 1,2 bytes structs on all platforms when ABI doesn't require it, so I made a change not to do it, so we process all small structs (1,2,3, 5, 6, 7) equally.  It gave me a small win on x64 Windows as well, I will do it for short and byte, but need to check what C++ compiler expects.

No changes when the feature is disabled (the old mode), expected changes on all platforms when enabled. Arm32/Arm64 is coming in another PR, there is non-trivial HFA logic there.

**There are many trivial commits, so I recommend reviewing by commits:**
4b9d166626a: Add a few const modifiers. **Small**

a642c9a3f51: Add more test cases. **No need to review**

84ba5848c7f: Initialize `ReturnTypeDesc` when we keep struct types. **Small**

07e41a3c098: Additional checks in `LowerRet`. **Small**

86151b4c91e: Support `return double(cnst int)`. **Small**

04f8959a477: Optimize `LowerRetStruct`: no need for bitcast when read from memory. **x64 windows diffs**

a0852bf518e: Prepare `LowerNode` for store local and local field to multireg. **Small**

79a16f06dff: Compile the new methods with FEATURE_MULTIREG_RET. **Small**

a4833e3141d: Improve `LowerRetStructLclVar`. 

Don't use bitcast if the source is in memory or has the same type.

dbf2072b7db: Extract `LowerStoreLocCommon`. **Refactoring only, no need to review**

a6bba379640: Support 3, 5. 6, 7 bytes structs in `LowerCallStruct`. **The main one**

Move call handling to the users.

93b31743596: Disable `JitDoOldStructRetyping` for x86 and x64. **Small**

Windows  x64 was supported in a previous PR, this adds x86 (Windows and Linux) and x64 Unix.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @CarolEidt @dotnet/jit-contrib 

jitstress, jitstress-regs have passed, diffs are similar to what we saw on x64 windows, arm will come in a separate PR after that one.

File: src/coreclr/src/jit/lower.cpp [hunk 1]
```diff
@@ -3233,16 +3270,30 @@ void Lowering::LowerRetStructLclVar(GenTreeUnOp* ret)
 // Arguments:
 //     call - The call node to lower.
 //
-// Note: it transforms the call's user.
+// Note: it skips multiReg return, it transforms the call's user for `GT_STOREIND`.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp:3273]: I would separate this into two comments (they're not really related). I would also restate the first as "This handles only single-register returns."

File: src/coreclr/src/jit/lower.cpp [hunk 2]
```diff
@@ -3252,39 +3303,99 @@ void Lowering::LowerCallStruct(GenTreeCall* call)
...
                 break;
 
             default:
                 unreached();
         }
     }
+#endif // !TARGET_ARMARCH
+}
+
+void Lowering::LowerStructBlockStore(GenTreeBlk* store)
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp:3334]: This needs a function header.

File: src/coreclr/src/jit/lower.cpp [hunk 3]
```diff
@@ -3252,39 +3303,99 @@ void Lowering::LowerCallStruct(GenTreeCall* call)
...
+//
+// Arguments:
+//     call - call with 3, 5, 6 or 7 return size that has to be spilled to memory.
+//
+// Return Value:
+//    load of the spilled variable.
+//
+GenTreeLclVar* Lowering::SpillStructCallResult(GenTreeCall* call) const
+{
+    // TODO-1stClassStructs: we can support that in codegen for `GT_STORE_BLK` without new temps.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp:3384]: Nit:
```suggestion
    // TODO-1stClassStructs: we can support this in codegen for `GT_STORE_BLK` without new temps.
```

**Review by CarolEidt — COMMENTED** [review]
Overall LGTM, with some minor comments & suggestions.

File: src/coreclr/src/jit/lower.cpp [hunk 4]
```diff
@@ -3031,16 +3040,44 @@ void Lowering::LowerStoreLocCommon(GenTreeLclVarCommon* lclStore)
...
+                        case 2:
+                        case 4:
+                        case 8:
+                        case 16:
+                            assert(regType != TYP_UNDEF);
+                            break;
+                        case 3:
+                        case 5:
+                        case 6:
+                        case 7:
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/lower.cpp:3060]: but in `3, 5, 6, 7` the assert is different: `assert(regType == TYP_UNDEF);`.
I can do:
```
                    unsigned size = layout->GetSize();
                    assert((size <= 8) || (size == 16));
                    bool isPowerOf2 = (((size - 1) & size) == 0);
                    bool isTypeDefined = (regType != TYP_UNDEF);
                    assert(isPowerOf2 == isTypeDefined);
```

File: src/coreclr/src/jit/lower.cpp [hunk 5]
```diff
@@ -255,6 +255,13 @@ GenTree* Lowering::LowerNode(GenTree* node)
 
         case GT_STORE_BLK:
         case GT_STORE_OBJ:
+            if (node->AsBlk()->Data()->IsCall())
+            {
+                assert(!comp->compDoOldStructRetyping());
+                LowerStructBlockStore(node->AsBlk());
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/lower.cpp:261]: Changed to `LowerStoreCallStruct`.

File: src/coreclr/src/jit/lower.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/coreclr/src/jit/lower.cpp]: Sorry I was confused! I think either is OK, but the code you show above makes it a little clearer.

**Review by CarolEidt — APPROVED** [review]
LGTM - thanks! Looking forward to the next one :-)

**Pull request merged**